syntax = "proto3";

import "envd/permissions/permissions.proto";

package envd.filesystem;

service FilesystemService {
    rpc Stat(StatRequest) returns (StatResponse);
    rpc MakeDir(MakeDirRequest) returns (MakeDirResponse);
    rpc List(ListRequest) returns (ListResponse);
    rpc Watch(WatchRequest) returns (stream WatchResponse);
    rpc Remove(RemoveRequest) returns (RemoveResponse);
}

message MakeDirRequest {
    string path = 1;

    permissions.User user = 2;
}

message MakeDirResponse {
    string path = 1;
}

message RemoveRequest {
    string path = 1;

    permissions.User user = 2;
}

message RemoveResponse {}

message StatRequest {
    string path = 1;

    permissions.User user = 2;
}

message StatResponse {
    EntryInfo entry = 1;
}

message EntryInfo {
    string name = 1;
    FileType type = 2;
}

enum FileType {
    FILE_TYPE_UNSPECIFIED = 0;
    FILE_TYPE_FILE = 1;
    FILE_TYPE_DIRECTORY = 2;
}

message ListRequest {
    string path = 1;

    permissions.User user = 2;
}

message ListResponse {
    repeated EntryInfo entries = 1;
}

message WatchRequest {
    string path = 1;

    permissions.User user = 2;
}

message WatchResponse {
    FilesystemEvent event = 1;
}

message FilesystemEvent {
    string path = 1;
    EventType type = 2;
}

enum EventType {
    EVENT_TYPE_UNSPECIFIED = 0;
    EVENT_TYPE_CREATE = 1;
    EVENT_TYPE_WRITE = 2;
    EVENT_TYPE_REMOVE = 3;
    EVENT_TYPE_RENAME = 4;
    EVENT_TYPE_CHMOD = 5;
}
